### PR TITLE
Add deleteTag endpoints in client APIs

### PR DESCRIPTION
**Pull request by ankit-db**
## What changes are proposed in this pull request?
 
Follow up PR on #1558 - add the deleteTag endpoints to all the user-facing API endpoints in MLFlow
 
## How is this patch tested?
 
Storage implementations are end to end tested, simple end to end tests have been added for delete tag to the user APIs here as well.
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
See #1558. 

### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [x] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by ankitmathur-db** [comment]
Still a WIP, but would like review on the sketch of how the endpoints are being added in.

File: mlflow/R/mlflow/R/tracking-runs.R [hunk 1]
```diff
@@ -188,7 +188,7 @@ mlflow_set_tag <- function(key, value, run_id = NULL, client = NULL) {
 
 #' Delete Tag
 #'
-#' Deletes a tag on a run. Tags are run metadata that can be updated during a run and
+#' Deletes a tag on a run. This is irreversible. Tags are run metadata that can be updated during a run and
 #'  after a run completes.
```
Comment (stbof) [inline mlflow/R/mlflow/R/tracking-runs.R]: Deletes a tag -> Delete a tag

File: docs/source/R-api.rst [hunk 2]
```diff
@@ -173,8 +173,8 @@ Arguments
 
 Delete Tag
 
-Deletes a tag on a run. Tags are run metadata that can be updated during
-a run and after a run completes.
+Deletes a tag on a run. This is irreversible. Tags are run metadata that
+can be updated during a run and after a run completes.
```
Comment (stbof) [inline docs/source/R-api.rst]: Deletes a tag -> Delete a tag

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowClient.java [hunk 3]
```diff
@@ -299,7 +299,7 @@ public void setTag(String runId, String key, String value) {
   }
 
   /**
-   * Delete a tag against the given run, given a String key.
+   * Delete a tag against the given run, given a String key. This is irreversible.
    * @param runId The ID of the run in which the tag exists to delete.
```
Comment (stbof) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowClient.java]: The ID of the run in which the tag exists to delete. -> ID of the run that contains the tag to delete

File: mlflow/store/file_store.py [hunk 4]
```diff
@@ -587,7 +587,7 @@ def set_tag(self, run_id, tag):
 
     def delete_tag(self, run_id, key):
         """
-        Delete a tag from a run.
+        Delete a tag from a run. This is irreversible.
         :param run_id: ID of the run that contains the tag to delete
         :param key: name of the tag to delete
```
Comment (stbof) [inline mlflow/store/file_store.py]: name of the tag to delete -> Name of the tag to delete

File: mlflow/store/rest_store.py [hunk 5]
```diff
@@ -191,7 +191,7 @@ def set_tag(self, run_id, tag):
 
     def delete_tag(self, run_id, key):
         """
-        Delete a tag for the specified run with the specified tag key.
+        Delete a tag for the specified run with the specified tag key. This is irreversible.
         :param run_id: String id for the run
         :param key: name of the tag to delete
```
Comment (stbof) [inline mlflow/store/rest_store.py]: String id for the run -> String ID of the run
name of the tag to delete -> Name of the tag to delete

File: mlflow/store/sqlalchemy_store.py [hunk 6]
```diff
@@ -465,7 +465,7 @@ def set_tag(self, run_id, tag):
 
     def delete_tag(self, run_id, key):
         """
-        Delete a tag from a run.
+        Delete a tag from a run. This is irreversible.
         :param run_id: ID of the run that contains the tag to delete
         :param key: name of the tag to delete
         """
```
Comment (stbof) [inline mlflow/store/sqlalchemy_store.py]: name of the tag to delete -> Name of the tag to delete

File: mlflow/tracking/client.py [hunk 7]
```diff
@@ -181,7 +181,9 @@ def set_tag(self, run_id, key, value):
 
     def delete_tag(self, run_id, key):
         """
-        Delete a tag on the run ID with a specific key.
+        :param run_id: Run ID of the run delete the tag from.
+        :param key: Name of the tag to delete.
```
Comment (stbof) [inline mlflow/tracking/client.py]: All the descriptions for run_id and key should be the same.

**Comment by ankitmathur-db** [comment]
Thanks for the review @stbof - I'll update the docs to reflect that feedback.

File: mlflow/java/client/src/test/java/org/mlflow/tracking/MlflowClientTest.java [hunk 8]
```diff
@@ -171,6 +171,39 @@ public void addGetRun() {
...
+    client.setTag(runId, "tag1", "val1");
+    client.deleteTag(runId, "tag0");
+    Run run = client.getRun(runId);
+    // test that the tag was correctly deleted.
+    for (RunTag rt : run.getData().getTagsList()) {
+      Assert.assertTrue(!rt.getKey().equals("tag0"));
+    }
+    // test that you can't delete a tag that doesn't already exist.
+    try {
+      client.deleteTag(runId, "fakeTag");
```
Comment (mparkhe) [inline mlflow/java/client/src/test/java/org/mlflow/tracking/MlflowClientTest.java]: would be interesting to test `deleteTag` on `"tag0"` (again) after confirming that it was deleted.

File: mlflow/store/file_store.py (hunk 4 above)
Comment (ankitmathur-db) [reply mlflow/store/file_store.py]: Resolved in previous commit

File: mlflow/store/sqlalchemy_store.py (hunk 6 above)
Comment (ankitmathur-db) [reply mlflow/store/sqlalchemy_store.py]: Resolved in previous PR

File: mlflow/java/client/src/test/java/org/mlflow/tracking/MlflowClientTest.java (hunk 8 above)
Comment (ankitmathur-db) [reply mlflow/java/client/src/test/java/org/mlflow/tracking/MlflowClientTest.java]: Added such a test.

File: mlflow/tracking/fluent.py [hunk 9]
```diff
@@ -187,6 +187,15 @@ def set_tag(key, value):
     MlflowClient().set_tag(run_id, key, value)
 
 
+def delete_tag(key):
```
Comment (dbczumar) [inline mlflow/tracking/fluent.py]: Attempting to invoke
```
import mlflow
mlflow.delete_tag(...)
```
yields
```
AttributeError: 'module' object has no attribute 'delete_tag'
```

Looks like we aren't exporting the `delete_tag` function to the top-level MLflow module. You can do that by adding an entry here: https://github.com/mlflow/mlflow/blob/9390f44e310faef7f030a75c6421754a4deff951/mlflow/__init__.py#L48

This also seems to imply that we don't have a test that calls `mlflow.delete_tag`; can we add one? `mlflow/tests/tracking/test_fluent.py` seems like a good location.

Comment (ankitmathur-db) [reply mlflow/tracking/fluent.py]: Thanks for flagging this - sorry about missing it. Next commit has a fix now.

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @ankitmathur-db!

**Pull request merged**